File: src/types/typedMetadata.ts
// ============================================================================
// Typed Metadata Interfaces
// TypeScript definitions matching Rust backend types
// ============================================================================

// ============================================================================
// File Metadata
// ============================================================================

export interface FileMetadata {
  fileTypeId?: string;
  fieldId?: string;
  difficulty?: number; // 1-5
  date?: string;
  solvedProoved?: boolean;
  solutionId?: string;
  bibliography?: string;
  fileContent?: string;
  preambleId?: string;
  buildCommand?: string;
  fileDescription?: string;
  // Arrays (junction tables)
  chapters?: string[];
  sections?: string[];
  subsections?: string[]; // NEW: 4th level
  exerciseTypes?: string[];
  requiredPackages?: string[];
  customTags?: string[];
  bibEntries?: string[];
}

// ============================================================================
// Document Metadata
// ============================================================================

export interface IncludedFile {
  fileId: string;
  orderIndex: number;
  filesDatabaseSource?: string;
  databaseType?: number;
}

export interface DocumentMetadata {
  title?: string;
  documentTypeId?: string;
  // Hierarchy fields (replacing basicFolder/subFolder/subsubFolder)
  fieldId?: string;
  chapters?: string[];
  sections?: string[];
  subsections?: string[];
  // Legacy folder fields (kept for backwards compatibility, auto-filled)
  basicFolder?: string;
  subFolder?: string;
  subsubFolder?: string;
  // Other fields
  date?: string;
  content?: string;
  preambleId?: string;
  buildCommand?: string;
  bibliography?: string;
  description?: string;
  solutionDocumentId?: string;
  // Arrays
  includedFiles?: IncludedFile[];
  customTags?: string[];
  bibEntries?: string[];
}

// ============================================================================
// Table Metadata
// ============================================================================

export interface TableMetadata {
  tableTypeId?: string;
  date?: string;
  content?: string;
  caption?: string;
  description?: string;

  // LaTeX specifics
  environment?: string; // tabular, tabularx, longtable, tabularray
  placement?: string;
  label?: string;
  width?: string;
  alignment?: string;

  // Dimensions
  rows?: number;
  columns?: number;

  // Arrays
  requiredPackages?: string[];
  customTags?: string[];
}

// ============================================================================
// Figure Metadata
// ============================================================================

export interface FigureMetadata {
  figureTypeId?: string;
  environment?: string;
  date?: string;
  content?: string;
  caption?: string;
  preambleId?: string;
  buildCommand?: string;
  description?: string;
  // Layout & Dimensions
  width?: string;
  height?: string;
  options?: string;
  tikzStyle?: string;
  label?: string;
  placement?: string;
  alignment?: string;
  // Arrays
  requiredPackages?: string[];
  customTags?: string[];
}

export interface FigureType {
  id: string;
  name: string;
  description?: string;
}

// ============================================================================
// Command Metadata
// ============================================================================

export interface CommandMetadata {
  name?: string;
  commandTypeId?: string;
  argumentsNum?: number;
  optionalArgument?: string;
  content?: string;
  example?: string;
  description?: string;
  builtIn?: boolean;
  // Arrays
  requiredPackages?: string[];
  customTags?: string[];
}

export interface CommandType {
  id: string;
  name: string;
  description?: string;
}

// ============================================================================
// Package Metadata
// ============================================================================

export interface PackageMetadata {
  name?: string;
  topicId?: string;
  date?: string;
  content?: string;
  description?: string;
  options?: string;
  builtIn?: boolean;
  documentation?: string;
  example?: string;
  // Arrays
  requiredPackages?: string[];
  topics?: string[];
  providedCommands?: string[];
  customTags?: string[];
}

// ============================================================================
// Preamble Metadata
// ============================================================================

export interface PreambleMetadata {
  name?: string;
  preambleTypeId?: string;
  content?: string;
  description?: string;
  builtIn?: boolean;
  // Enriched fields
  engines?: string;
  date?: string;
  className?: string; // Mapped from 'class' in DB
  paperSize?: string;
  fontSize?: number;
  options?: string;
  languages?: string;
  geometry?: string;
  author?: string;
  title?: string;
  // Booleans
  useBibliography?: boolean;
  bibCompileEngine?: string;
  makeIndex?: boolean;
  makeGlossaries?: boolean;
  hasToc?: boolean;
  hasLot?: boolean;
  hasLof?: boolean;
  // Arrays
  requiredPackages?: string[];
  commandTypes?: string[];
  providedCommands?: string[];
}

// ============================================================================
// Class Metadata
// ============================================================================

export interface ClassMetadata {
  name?: string;
  fileTypeId?: string;
  date?: string;
  content?: string;
  description?: string;
  // Enriched fields
  engines?: string;
  paperSize?: string;
  fontSize?: number;
  geometry?: string;
  options?: string;
  languages?: string;
  // Arrays
  customTags?: string[];
  requiredPackages?: string[];
  providedCommands?: string[];
}

// ============================================================================
// Bibliography Metadata
// ============================================================================

export interface BibliographyPerson {
  role: "author" | "editor" | "translator";
  fullName: string;
}

export interface BibliographyMetadata {
  entryType?: string; // article, book, etc.
  citationKey?: string;

  // Periodic / Series
  journal?: string;
  volume?: string;
  series?: string;
  number?: string;
  issue?: string;

  // Date
  year?: string;
  month?: string;

  // Publication
  publisher?: string;
  edition?: string;
  institution?: string;
  school?: string;
  organization?: string;
  address?: string;
  location?: string;

  // Identification
  isbn?: string;
  issn?: string;
  doi?: string;
  url?: string;
  language?: string;

  // Content
  title?: string;
  subtitle?: string;
  booktitle?: string;
  chapter?: string;
  pages?: string;
  abstract?: string;
  note?: string; // Description/Note
  crossref?: string;

  // Lists
  authors?: string[]; // Simplified for UI, mapped to persons in DB
  editors?: string[];
  translators?: string[];

  // Extras
  extras?: Record<string, string>; // Custom fields
}

// ============================================================================
// Unified TypedMetadata Union
// ============================================================================

export type TypedMetadata =
  | { type: "file"; data: FileMetadata }
  | { type: "document"; data: DocumentMetadata }
  | { type: "bibliography"; data: BibliographyMetadata } // Added bibliography
  | { type: "table"; data: TableMetadata }
  | { type: "figure"; data: FigureMetadata }
  | { type: "command"; data: CommandMetadata }
  | { type: "package"; data: PackageMetadata }
  | { type: "preamble"; data: PreambleMetadata }
  | { type: "class"; data: ClassMetadata };

// ============================================================================
// Lookup Data Types
// ============================================================================

export interface Field {
  id: string;
  name: string;
  description?: string;
}

export interface Chapter {
  id: string;
  name: string;
  fieldId: string;
}

export interface Section {
  id: string;
  name: string;
  chapterId: string;
}

export interface Subsection {
  id: string;
  name: string;
  sectionId: string;
}

export interface ExerciseType {
  id: string;
  name: string;
  description?: string;
}

export interface FileType {
  id: string;
  name: string;
  folderName?: string;
  solvable?: boolean;
  description?: string;
}

export interface DocumentType {
  id: string;
  name: string;
  description?: string;
}

export interface TableType {
  id: string;
  name: string;
  description?: string;
}

export interface CustomTag {
  tag: string;
  createdAt?: string;
}

export interface PreambleType {
  id: string;
  name: string;
  description?: string;
}

export interface TeXLivePackage {
  id: string;
  description?: string;
}

export interface PackageTopic {
  id: string;
  name: string;
  description?: string;
}

export interface MacroCommandType {
  id: string;
  name: string;
  description?: string;
}

// ============================================================================
// Resource with Typed Metadata
// ============================================================================

export interface ResourceWithTypedMetadata {
  id: string;
  path: string;
  type: string;
  collection: string;
  title?: string;
  contentHash?: string;
  createdAt: string;
  updatedAt: string;
  typedMetadata?: TypedMetadata;
}

// ============================================================================
// Helper Types
// ============================================================================

export type ResourceType =
  | "file"
  | "document"
  | "bibliography"
  | "table"
  | "figure"
  | "command"
  | "package"
  | "preamble"
  | "class";

export type MetadataByType<T extends ResourceType> = T extends "file"
  ? FileMetadata
  : T extends "document"
  ? DocumentMetadata
  : T extends "bibliography"
  ? BibliographyMetadata
  : T extends "table"
  ? TableMetadata
  : T extends "figure"
  ? FigureMetadata
  : T extends "command"
  ? CommandMetadata
  : T extends "package"
  ? PackageMetadata
  : T extends "preamble"
  ? PreambleMetadata
  : T extends "class"
  ? ClassMetadata
  : never;

File: src-tauri/src/db/typed_metadata.rs
// ============================================================================
// Typed Metadata Database Operations
// CRUD operations for strongly-typed metadata
// ============================================================================

use crate::types::metadata::*;
use rusqlite::{params, Connection, OptionalExtension, Result};

// ============================================================================
// CREATE Operations
// ============================================================================

/// Save FileMetadata to database
pub fn save_file_metadata(
    conn: &Connection,
    resource_id: &str,
    metadata: &FileMetadata,
) -> Result<()> {
    // Insert main record
    conn.execute(
        "INSERT OR REPLACE INTO resource_files (
            resource_id, file_type_id, field_id, difficulty, date,
            solved_prooved, solution_id, bibliography, file_content,
            preamble_id, build_command, file_description
        ) VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11, ?12)",
        params![
            resource_id,
            metadata.file_type_id,
            metadata.field_id,
            metadata.difficulty,
            metadata.date,
            metadata.solved_prooved,
            metadata.solution_id,
            metadata.bibliography,
            metadata.file_content,
            metadata.preamble_id,
            metadata.build_command,
            metadata.file_description,
        ],
    )?;

    // Save chapters
    if let Some(chapters) = &metadata.chapters {
        conn.execute(
            "DELETE FROM resource_file_chapters WHERE resource_id = ?1",
            params![resource_id],
        )?;
        for chapter_id in chapters {
            conn.execute(
                "INSERT OR IGNORE INTO resource_file_chapters (resource_id, chapter_id) VALUES (?1, ?2)",
                params![resource_id, chapter_id],
            )?;
        }
    }

    // Save sections
    if let Some(sections) = &metadata.sections {
        conn.execute(
            "DELETE FROM resource_file_sections WHERE resource_id = ?1",
            params![resource_id],
        )?;
        for section_id in sections {
            conn.execute(
                "INSERT OR IGNORE INTO resource_file_sections (resource_id, section_id) VALUES (?1, ?2)",
                params![resource_id, section_id],
            )?;
        }
    }

    // Save exercise types
    if let Some(exercise_types) = &metadata.exercise_types {
        conn.execute(
            "DELETE FROM resource_file_exercise_types WHERE resource_id = ?1",
            params![resource_id],
        )?;
        for exercise_type_id in exercise_types {
            conn.execute(
                "INSERT OR IGNORE INTO resource_file_exercise_types (resource_id, exercise_type_id) VALUES (?1, ?2)",
                params![resource_id, exercise_type_id],
            )?;
        }
    }

    // Save custom tags
    if let Some(tags) = &metadata.custom_tags {
        conn.execute(
            "DELETE FROM resource_file_tags WHERE resource_id = ?1",
            params![resource_id],
        )?;
        for tag in tags {
            conn.execute(
                "INSERT OR IGNORE INTO custom_tags (tag) VALUES (?1)",
                params![tag],
            )?;
            conn.execute(
                "INSERT OR IGNORE INTO resource_file_tags (resource_id, tag) VALUES (?1, ?2)",
                params![resource_id, tag],
            )?;
        }
    }

    Ok(())
}

/// Save DocumentMetadata to database
pub fn save_document_metadata(
    conn: &Connection,
    resource_id: &str,
    metadata: &DocumentMetadata,
) -> Result<()> {
    conn.execute(
        "INSERT OR REPLACE INTO resource_documents (
            resource_id, title, document_type_id, basic_folder, sub_folder,
            subsub_folder, date, content, preamble_id, build_command,
            needs_update, bibliography, description, solution_document_id
        ) VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11, ?12, ?13, ?14)",
        params![
            resource_id,
            metadata.title,
            metadata.document_type_id,
            metadata.basic_folder.as_deref().unwrap_or("General"),
            metadata.sub_folder.as_deref().unwrap_or("General"),
            metadata.subsub_folder,
            metadata.date,
            metadata.content,
            metadata.preamble_id,
            metadata.build_command,
            metadata.needs_update,
            metadata.bibliography,
            metadata.description,
            metadata.solution_document_id,
        ],
    )?;

    // Save custom tags
    if let Some(tags) = &metadata.custom_tags {
        conn.execute(
            "DELETE FROM resource_document_tags WHERE resource_id = ?1",
            params![resource_id],
        )?;
        for tag in tags {
            conn.execute(
                "INSERT OR IGNORE INTO custom_tags (tag) VALUES (?1)",
                params![tag],
            )?;
            conn.execute(
                "INSERT OR IGNORE INTO resource_document_tags (resource_id, tag) VALUES (?1, ?2)",
                params![resource_id, tag],
            )?;
        }
    }

    Ok(())
}

// ============================================================================
// READ Operations
// ============================================================================

/// Load FileMetadata from database
pub fn load_file_metadata(conn: &Connection, resource_id: &str) -> Result<Option<FileMetadata>> {
    let mut stmt = conn.prepare(
        "SELECT file_type_id, field_id, difficulty, date, solved_prooved,
                solution_id, bibliography, file_content, preamble_id,
                build_command, file_description
         FROM resource_files WHERE resource_id = ?1",
    )?;

    let metadata = stmt
        .query_row(params![resource_id], |row| {
            Ok(FileMetadata {
                file_type_id: row.get(0)?,
                field_id: row.get(1)?,
                difficulty: row.get(2)?,
                date: row.get(3)?,
                solved_prooved: row.get(4)?,
                solution_id: row.get(5)?,
                bibliography: row.get(6)?,
                file_content: row.get(7)?,
                preamble_id: row.get(8)?,
                build_command: row.get(9)?,
                file_description: row.get(10)?,
                chapters: None,
                sections: None,
                exercise_types: None,
                required_packages: None,
                custom_tags: None,
                bib_entries: None,
            })
        })
        .optional()?;

    if let Some(mut meta) = metadata {
        // Load chapters
        let chapters: Vec<String> = conn
            .prepare("SELECT chapter_id FROM resource_file_chapters WHERE resource_id = ?1")?
            .query_map(params![resource_id], |row| row.get(0))?
            .collect::<Result<Vec<String>>>()?;
        if !chapters.is_empty() {
            meta.chapters = Some(chapters);
        }

        // Load sections
        let sections: Vec<String> = conn
            .prepare("SELECT section_id FROM resource_file_sections WHERE resource_id = ?1")?
            .query_map(params![resource_id], |row| row.get(0))?
            .collect::<Result<Vec<String>>>()?;
        if !sections.is_empty() {
            meta.sections = Some(sections);
        }

        // Load exercise types
        let exercise_types: Vec<String> = conn
            .prepare(
                "SELECT exercise_type_id FROM resource_file_exercise_types WHERE resource_id = ?1",
            )?
            .query_map(params![resource_id], |row| row.get(0))?
            .collect::<Result<Vec<String>>>()?;
        if !exercise_types.is_empty() {
            meta.exercise_types = Some(exercise_types);
        }

        // Load custom tags
        let tags: Vec<String> = conn
            .prepare("SELECT tag FROM resource_file_tags WHERE resource_id = ?1")?
            .query_map(params![resource_id], |row| row.get(0))?
            .collect::<Result<Vec<String>>>()?;
        if !tags.is_empty() {
            meta.custom_tags = Some(tags);
        }

        return Ok(Some(meta));
    }

    Ok(None)
}

/// Load DocumentMetadata from database
pub fn load_document_metadata(
    conn: &Connection,
    resource_id: &str,
) -> Result<Option<DocumentMetadata>> {
    let mut stmt = conn.prepare(
        "SELECT title, document_type_id, basic_folder, sub_folder, subsub_folder,
                date, content, preamble_id, build_command, needs_update,
                bibliography, description, solution_document_id
         FROM resource_documents WHERE resource_id = ?1",
    )?;

    let metadata = stmt
        .query_row(params![resource_id], |row| {
            Ok(DocumentMetadata {
                title: row.get(0)?,
                document_type_id: row.get(1)?,
                basic_folder: row.get(2)?,
                sub_folder: row.get(3)?,
                subsub_folder: row.get(4)?,
                date: row.get(5)?,
                content: row.get(6)?,
                preamble_id: row.get(7)?,
                build_command: row.get(8)?,
                needs_update: row.get(9)?,
                bibliography: row.get(10)?,
                description: row.get(11)?,
                solution_document_id: row.get(12)?,
                included_files: None,
                custom_tags: None,
                bib_entries: None,
            })
        })
        .optional()?;

    if let Some(mut meta) = metadata {
        // Load custom tags
        let tags: Vec<String> = conn
            .prepare("SELECT tag FROM resource_document_tags WHERE resource_id = ?1")?
            .query_map(params![resource_id], |row| row.get(0))?
            .collect::<Result<Vec<String>>>()?;
        if !tags.is_empty() {
            meta.custom_tags = Some(tags);
        }

        return Ok(Some(meta));
    }

    Ok(None)
}

// ============================================================================
// Generic load function by resource type
// ============================================================================

pub fn load_typed_metadata(
    conn: &Connection,
    resource_id: &str,
    resource_type: &str,
) -> Result<Option<TypedMetadata>> {
    match resource_type {
        "file" => {
            if let Some(meta) = load_file_metadata(conn, resource_id)? {
                Ok(Some(TypedMetadata::File(meta)))
            } else {
                Ok(None)
            }
        }
        "document" => {
            if let Some(meta) = load_document_metadata(conn, resource_id)? {
                Ok(Some(TypedMetadata::Document(meta)))
            } else {
                Ok(None)
            }
        }
        _ => Ok(None),
    }
}
